File: android/src/main/kotlin/com/example/growth_auth0/GrowthAuth0Plugin.kt
package com.example.growth_auth0

import Auth
import android.app.Activity
import android.content.Context
import android.util.Log
import androidx.annotation.NonNull
import com.auth0.android.authentication.AuthenticationException
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch

/** GrowthAuth0Plugin */
class GrowthAuth0Plugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel

    private lateinit var context: Context

    private var activity: Activity? = null

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "growth_auth0")
        channel.setMethodCallHandler(this)
        context = flutterPluginBinding.applicationContext
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: MethodChannel.Result) {
        Log.d("Auth0Plugin", call.method);
        if (call.method == "initAuth") {

            val authClientId: String = call.argument("authClientId") ?: ""
            val authDomain = call.argument("authDomain") ?: ""

            Auth.initAuth(context, authClientId, authDomain)
            result.success(null)

        } else if (call.method == "loginWithUniversal") {
            if (activity == null) {
                result.error("ACTIVITY_NOT_AVAILABLE", "Activity is not available.", null)
                return
            }

            GlobalScope.launch { handleLoginWithUniversalAsync(call, result, activity!!) }
        } else if (call.method == "login") {
            GlobalScope.launch { handleLogin(call, result) }
        } else if (call.method == "passwordLessWithEmail") {
            GlobalScope.launch { handlePasswordLessWithEmail(call, result) }

        } else if (call.method == "passwordLessWithSMS") {
            GlobalScope.launch { handlePasswordLessWithSMS(call, result) }

        } else if (call.method == "loginWithEmail") {
            GlobalScope.launch { handleLoginWithEmail(call, result) }

        } else if (call.method == "loginWithPhoneNumber") {
            GlobalScope.launch { handleLoginWithPhoneNumber(call, result) }

        } else if (call.method == "checkIsLogged") {
            result.success(Auth.checkIsLogged())

        } else if (call.method == "getAccessToken") {
            GlobalScope.launch { handleGetToken(call, result) }

        } else if (call.method == "getUserInfo") {
            GlobalScope.launch { handleGetUserInfo(call, result) }

        } else if (call.method == "logout") {
            handleLogout(result)

        } else {
            result.notImplemented()
        }
    }


    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    private suspend fun handleLoginWithUniversalAsync(@NonNull call: MethodCall, @NonNull result: MethodChannel.Result, @NonNull context: Context) {
        val audience = call.argument("audience") ?: ""
        val scope = call.argument("scope") ?: ""
        val scheme = call.argument("scheme") ?: ""

        try {
            result.success(Auth.loginWithUniversalAsync(context, audience, scope, scheme).await())
        } catch (e: AuthenticationException) {
            result.error("Auth0Plugin AuthenticationException handle loginWith universal async", e.getDescription(), null)
        } catch (e: Throwable) {
            result.error("Auth0Plugin Error handle loginWith universal async", e.message, null)
        }
    }

    private suspend fun handleLogin(@NonNull call: MethodCall, @NonNull result: MethodChannel.Result) {
        val email = call.argument("email") ?: ""
        val password = call.argument("password") ?: ""
        val realmOrConnection = call.argument("realmOrConnection") ?: ""
        val audience = call.argument("audience") ?: ""
        val scope = call.argument("scope") ?: ""

        try {
            result.success(Auth.loginAsync(email, password, realmOrConnection, audience, scope).await())
        } catch (e: AuthenticationException) {
            result.error("Auth0Plugin AuthenticationException handle login", e.getDescription(), null)
        } catch (e: Throwable) {
            result.error("Auth0Plugin Error handle login", e.message, null)
        }
    }

    private suspend fun handlePasswordLessWithEmail(@NonNull call: MethodCall, @NonNull result: MethodChannel.Result) {
        val email = call.argument("email") ?: ""

        try {
            result.success(Auth.passwordLessWithEmailAsync(email).await())
        } catch (e: AuthenticationException) {
            result.error("Auth0Plugin AuthenticationException handle email code", e.getDescription(), null)
        } catch (e: Throwable) {
            result.error("Auth0Plugin Error handle email code", e.message, null)
        }
    }

    private suspend fun handlePasswordLessWithSMS(@NonNull call: MethodCall, @NonNull result: MethodChannel.Result) {
        val phoneNumber = call.argument("phoneNumber") ?: ""

        try {
            result.success(Auth.passwordLessWithSMSAsync(phoneNumber).await())
        } catch (e: AuthenticationException) {
            result.error("Auth0Plugin AuthenticationException handle phone number", e.getDescription(), null)
        } catch (e: Throwable) {
            result.error("Auth0Plugin Error handle phone number", e.message, null)
        }

    }

    private suspend fun handleLoginWithEmail(@NonNull call: MethodCall, @NonNull result: MethodChannel.Result) {
        val email = call.argument("email") ?: ""
        val code = call.argument("code") ?: ""
        val audience = call.argument("audience") ?: ""
        val scope = call.argument("scope") ?: ""

        try {
            result.success(Auth.loginWithEmailAsync(email, code, audience, scope).await())
        } catch (e: Throwable) {
            result.error("Auth0Plugin Error handle code", e.message, null)
        }

    }

    private suspend fun handleLoginWithPhoneNumber(@NonNull call: MethodCall, @NonNull result: MethodChannel.Result) {
        val phoneNumber = call.argument("phoneNumber") ?: ""
        val code = call.argument("code") ?: ""
        val audience = call.argument("audience") ?: ""
        val scope = call.argument("scope") ?: ""

        try {
            result.success(Auth.loginWithPhoneNumberAsync(phoneNumber, code, audience, scope).await())
        } catch (e: Throwable) {
            result.error("Auth0Plugin Error handle code phone", e.message, null)
        }

    }

    private suspend fun handleGetToken(@NonNull call: MethodCall, @NonNull result: MethodChannel.Result) {
        try {
            result.success(Auth.getAccessTokenAsync().await())
        } catch (e: Throwable) {
            result.error("Auth0Plugin Error handle get accessToken", e.message, null)
        }

    }

    private suspend fun handleGetUserInfo(@NonNull call: MethodCall, @NonNull result: MethodChannel.Result) {
        try {

            val accessToken = call.argument("accessToken") ?: ""

            result.success(Auth.getUserInfoAsync(accessToken).await())
        } catch (e: Throwable) {
            result.error("Auth0Plugin Error handle get user ifno", e.message, null)
        }

    }

    private fun handleLogout(@NonNull result: MethodChannel.Result) {
        try {
            Auth.logout()
            result.success(true)
        } catch (e: Throwable) {
            result.error("Auth0Plugin Error logout", e.message, null)
        }
    }
}

File: ios/Classes/GrowthAuth0Plugin.swift
import Flutter
import UIKit

public class GrowthAuth0Plugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "growth_auth0", binaryMessenger: registrar.messenger())
        let instance = GrowthAuth0Plugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch(call.method) {
            case "initAuth":
                handleInit(call: call, result: result)
                
        case  "login":
                handleLogin(call: call, result: result)
                break;
            
        case  "passwordLessWithEmail":
                handlePasswordLessWithEmail(call: call, result: result)
                break;
            
            case "passwordLessWithSMS":
                handlePasswordLessWithSMS(call: call, result: result)
                
            case "loginWithEmail":
                handleLoginWithEmail(call: call, result: result)
                break;
            
                break;
            case "loginWithPhoneNumber":
                handleLoginWithPhoneNumber(call: call, result: result)
                break;
                
            case "checkIsLogged":
                result(Auth.shared.checkIsLogged())
                break;
                
            case "getAccessToken":
                handleGetAccessToken(result: result)
                break;
            
            case "clearCredentials":
                result(Auth.shared.clearCredentials())
                break;
            
            case "logout":
                result(Auth.shared.logout())
                break;
            default:
                result(getBaseError())
        }
    
        
    }
    
    private func handleInit(call: FlutterMethodCall, result: @escaping FlutterResult) {
    
        if let myArgs = call.arguments as? [String: Any],
           let authClientId =  myArgs["authClientId"] as? String,
           let authDomain =  myArgs["authDomain"] as? String {
           
            Auth.shared.initAuth(clientId: authClientId, domain: authDomain)
            debugPrint("Auth0Plugin initAuth complete")
            
            result(true)
            
        } else {
            debugPrint("Auth0Plugin error")
            result(getBaseError())
        }
    }
    
    private func handlePasswordLessWithEmail(call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let myArgs = call.arguments as? [String: Any],
           let email =  myArgs["email"] as? String {
            
            //Result<Void, AuthenticationError>
            Auth.shared.passwordLessWithEmail(
                email: email,
                onSuccess: {
                    (isSuccess: Bool) in result(isSuccess)
                },
                onError: {
                    (error: FlutterError) in result(error)
                }
            )
            
        } else {
            result(getBaseError())
        }
    }
    
    private func handlePasswordLessWithSMS(call: FlutterMethodCall, result: @escaping FlutterResult) {
        
        if let myArgs = call.arguments as? [String: Any],
           let phoneNumber =  myArgs["phoneNumber"] as? String {
            
            //Result<Void, AuthenticationError>
            Auth.shared.passwordLessWithSMS(
                phoneNumber: phoneNumber,
                onSuccess: {
                    (isSuccess: Bool) in result(isSuccess)
                },
                onError: {
                    (error: FlutterError) in result(error)
                }
            )
            
        } else {
            result(getBaseError())
        }
    }
    
    private func handleLogin(call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let myArgs = call.arguments as? [String: Any],
           let email =  myArgs["email"] as? String,
           let password =  myArgs["password"] as? String,
           let realmOrConnection =  myArgs["realmOrConnection"] as? String {
            
            var audience: String? = nil
            var scope: String? = nil
            
            if(myArgs["audience"] is String?) {
                audience = myArgs["audience"] as! String?
            }
            
            if(myArgs["scope"] is String?) {
                scope = myArgs["scope"] as! String?
            }
            
            Auth.shared.login(
                email: email,
                password: password,
                audience: audience,
                scope: scope,
                realmOrConnection: realmOrConnection,
                onSuccess: {
                    (isSuccess: Bool) in result(isSuccess)
                },
                onError: {
                    (error: FlutterError) in result(error)
                }
            )
        } else {
            result(getBaseError())
        }
    }
    
    private func handleLoginWithEmail(call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let myArgs = call.arguments as? [String: Any],
           let email =  myArgs["email"] as? String,
           let code =  myArgs["code"] as? String {
            
            var audience: String? = nil
            var scope: String? = nil
            
            if(myArgs["audience"] is String?) {
                audience = myArgs["audience"] as! String?
            }
            
            if(myArgs["scope"] is String?) {
                scope = myArgs["scope"] as! String?
            }
            
            Auth.shared.loginWithEmail(
                email: email,
                code: code,
                audience: audience,
                scope: scope,
                onSuccess: {
                    (isSuccess: Bool) in result(isSuccess)
                },
                onError: {
                    (error: FlutterError) in result(error)
                }
            )
            
        } else {
            result(getBaseError())
        }
    }
    
    private func handleLoginWithPhoneNumber(call: FlutterMethodCall, result: @escaping FlutterResult) {
        
        if let myArgs = call.arguments as? [String: Any],
           let phoneNumber =  myArgs["phoneNumber"] as? String,
           let code =  myArgs["code"] as? String {
            
            var audience: String?
            var scope: String?
            
            if(myArgs["audience"] is String?) {
                audience = myArgs["audience"] as! String?
            }
            
            if(myArgs["scope"] is String?) {
                scope = myArgs["scope"] as! String?
            }
            
            Auth.shared.loginWithPhoneNumber(
                phoneNumber: phoneNumber,
                code: code,
                audience: audience,
                scope: scope,
                onSuccess: {
                    (isSuccess: Bool) in result(isSuccess)
                },
                onError: {
                    (error: FlutterError) in result(error)
                }
            )
            
        } else {
            result(getBaseError())
        }
    }
    
    private func handleGetAccessToken(result: @escaping FlutterResult) {
        Auth.shared.getAccessToken(
            onSuccess: {
                (accessToken: String) in result(accessToken)
            },
            onError: {
                (error: FlutterError) in result(error)
            }
        )
    }
    
    private func getBaseError() -> FlutterError {
       return FlutterError(code: "Auth0Plugin error", message: "", details: "")
    }
    
    
}
